nitro-protocol/contracts/InterestBearingApp.sol

```diff
@@ -0,0 +1,150 @@
+
+// InterestBearingApp is a ForceMoveApp that allows a lender to earn interest
+// on a deposit. It functions as a ConsensusApp with the following additional rule:
+//   - the lender can unilaterally transition from state n to state n+1,
+//     forcing calculated interest into the lender's Outcome allocation
+contract InterestBearingApp is IForceMoveApp {
+    struct Funds {
+        address[] asset;
+        uint256[] amount;
+    }
+
+    struct InterestAppData {
+        // the per-block interest rate, expressed as the denominator of a fraction
+        // eg, 1% per block would be expressed by interestPerBlockDivisor = 100,
+        //     2% per block would be expressed by interestPerBlockDivisor = 50,
+        //     0.1% per block would be expressed by interestPerBlockDivisor = 1000,
+        // etc.
+        uint256 interestPerBlockDivisor;
+        // the block number of the latest principal adjustment
+        uint256 blocknumber;
+        // the current principal. Decreases as the borrower earns via the channel.
+        Funds principal;
+        // the total interest collected so far. Strictly increasing.
+        // The value of the lender's allocation can never be less than this:
+        // ie, when the lender's collectedInterest grows to be equal to their
+        // allocation, the channel is effectively spent and can be concluded.
+        Funds collectedInterest;
+    }
+
+    enum AllocationIndicies {
+        borrower, // intends to earn service fees up to a limit of the financier's deposit
+        lender // makes initial deposit and earns interest
+    }
+
+    function requireStateSupported(
+        FixedPart calldata,
+        RecoveredVariablePart[] calldata proof,
+        RecoveredVariablePart calldata candidate
+    ) external view override {
+        if (proof.length == 0) {
+            // unanimous consensus check
+            require(
+                NitroUtils.getClaimedSignersNum(candidate.signedBy) == 2,
+                '!unanimous; |proof|=0'
+            );
+            return;
+        } else if (proof.length == 1) {
+            // check that proof[0] -> candidate respects the stated interest rate.
+            // Requires:
+            //  - proof state is unanimous
+            //  - candidate state immediately follows proof state (by turnNum)
+            //  - the lender has not taken more funds than owed according
+            //    to the interest rate agreement of the channel
+            require(
+                NitroUtils.getClaimedSignersNum(proof[0].signedBy) == 2,
+                '!unanimous proof state'
+            );
+            require(
+                proof[0].variablePart.turnNum + 1 == candidate.variablePart.turnNum,
+                'turn(candidate) != turn(proof)+1'
+            );
+
+            Funds memory outstandingInterest = computeOutstandingInterest(
+                abi.decode(proof[0].variablePart.appData, (InterestAppData))
```

Comment:
Is there anything stopping the `lender` from changing the `interestPerBlockDivisor` in the `candidate`, effectively letting them charge whatever interest rate they want? I think we'd want to enforce here that the `interestPerBlockDivisor` hasn't changed from `proof->candidate`?  The `interestPerBlockDivisor` can still be changed co-operatively by the borrower and lender.

Reply:
Ah good point, I missed that we're using the app data from the proof `proof[0].variablePart.appData`